### DISCUSSION TITLE
Scripting epoch time calculation

### DISCUSSION THREAD

**Discussion by Andi-Coder** [arendst/Tasmota#11285] [Ideas]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._
The epoch variable at scripting section will only calculate every 8 second.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [X] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [X] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [X] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [X] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [X] Device used (e.g., Sonoff Basic): ESP8266EX
- [ ] Tasmota binary firmware version number used: 8.3.1
  - [ ] Pre-compiled
  - [X] Self-compiled
- [X] Flashing tools used: esptool
- [X] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
11:08:53 CMD: Backlog Template; Module; GPIO 255
11:08:53 MQT: stat/XX/RESULT = {"NAME":"Tuya Thermostat","GPIO":[255,255,255,255,255,255,0,0,255,255,255,255,255],"FLAG":15,"BASE":54}
11:08:54 MQT: stat/XX/RESULT = {"Module":{"0":"Tuya Thermostat"}}
11:08:54 MQT: stat/XX/RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"107":"Tuya Tx"},"GPIO2":{"0":"None"},"GPIO3":{"108":"Tuya Rx"},"GPIO4":{"0":"None"},"GPIO5":{"0":"None"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"0":"None"},"GPIO13":{"0":"None"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"}}



```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [X] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
11:10:27 CMD: Status 0
11:10:27 MQT: stat/XX/STATUS = {"Status":{"Module":0,"DeviceName":"XX","FriendlyName":["XX"],"Topic":"XX","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0}}
11:10:27 MQT: stat/XX/STATUS1 = {"StatusPRM":{"Baudrate":9600,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://thehackbox.org/tasmota/release/tasmota.bin","RestartReason":"Software/System restart","Uptime":"0T00:44:12","StartupUTC":"2021-03-10T09:26:15","Sleep":50,"CfgHolder":4617,"BootCount":71,"BCResetTime":"2021-01-10T12:48:39","SaveCount":491,"SaveAddress":"1F5000"}}
11:10:27 MQT: stat/XX/STATUS2 = {"StatusFWR":{"Version":"8.3.1(tasmota)","BuildDateTime":"2020-08-02T12:51:48","Boot":31,"Core":"2_7_2","SDK":"2.2.2-dev(38a443e)","Hardware":"ESP8266EX","CR":"414/699"}}
11:10:27 MQT: stat/XX/STATUS3 = {"StatusLOG":{"SerialLog":0,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["XX",""],"TelePeriod":60,"Resolution":"558180C0","SetOption":["00008009","2805C8000100060000005A00000000000000","10000000","00000000"]}}
11:10:27 MQT: stat/XX/STATUS4 = {"StatusMEM":{"ProgramSize":562,"Free":1464,"Heap":23,"ProgramFlashSize":2048,"FlashSize":2048,"FlashChipId":"1540C8","FlashMode":2,"Features":["00000809","07802796","000983A1","00000404","010000C0","00000900","00000020"],"Drivers":"1,2,3,4,8,9,10,16,17,20,27,29","Sensors":"2,3,9"}}
11:10:27 MQT: stat/XX/STATUS5 = {"StatusNET":{"Hostname":"XX","IPAddress":"192.168.XX.143","Gateway":"192.168.XX.1","Subnetmask":"255.255.255.0","DNSServer":"192.168.XX.1","Mac":"98:F4:AB:EF:8D:E0","Webserver":2,"WifiConfig":4,"WifiPower":17.0}}
11:10:27 MQT: stat/XX/STATUS6 = {"StatusMQT":{"MqttHost":"192.168.XX.100","MqttPort":1883,"MqttClientMask":"XX","MqttClient":"XX","MqttUser":"DVES_USER","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30}}
11:10:27 MQT: stat/XX/STATUS7 = {"StatusTIM":{"UTC":"2021-03-10T10:10:27","Local":"2021-03-10T11:10:27","StartDST":"2021-03-28T02:00:00","EndDST":"2021-10-31T03:00:00","Timezone":99,"Sunrise":"06:48","Sunset":"18:18"}}
11:10:27 MQT: stat/XX/STATUS10 = {"StatusSNS":{"Time":"2021-03-10T11:10:27"}}
11:10:27 MQT: stat/XX/STATUS11 = {"StatusSTS":{"Time":"2021-03-10T11:10:27","Uptime":"0T00:44:12","UptimeSec":2652,"Heap":23,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"OFF","Wifi":{"AP":1,"SSId":"HOTSPOT","BSSId":"F0:B0:14:D2:98:F5","Channel":11,"RSSI":86,"Signal":-57,"LinkCount":1,"Downtime":"0T00:00:05"}}}


```
- [X] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:
11:08:55 69070136.00
11:08:56 69070136.00
11:08:57 69070136.00
11:08:58 69070136.00
11:08:59 69070136.00
11:09:00 69070144.00
11:09:01 69070144.00
11:09:02 69070144.00
11:09:03 69070144.00
11:09:04 69070144.00
11:09:05 69070144.00
11:09:06 69070144.00
11:09:07 69070144.00
11:09:08 69070144.00
11:09:09 69070152.00
11:09:10 69070152.00
11:09:11 69070152.00
11:09:12 69070152.00
11:09:13 69070152.00
11:09:14 69070152.00
11:09:55 69070192.00
11:09:56 69070192.00
11:09:57 69070200.00
11:09:58 69070200.00
11:09:59 69070200.00
11:10:00 69070200.00
11:10:01 69070200.00
11:10:02 69070200.00
11:10:03 69070200.00
11:10:04 69070208.00
11:10:05 69070208.00
11:10:06 69070208.00
11:10:07 69070208.00
11:10:08 69070208.00
11:10:09 69070208.00
11:10:10 69070208.00
11:10:11 69070208.00
11:10:12 69070208.00
11:10:13 69070216.00
11:10:14 69070216.00
11:10:15 69070216.00
11:10:16 69070216.00
11:10:17 69070216.00
11:10:18 69070216.00
11:10:19 69070216.00
11:10:20 69070224.00
11:10:21 69070224.00
11:10:22 69070224.00
11:10:23 69070224.00
11:10:24 69070224.00
11:10:25 69070224.00

```

### TO REPRODUCE
_Steps to reproduce the behavior:_
At Scripting Section:
>D
...
>S
print %epoch%

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._
Expected that epoch change every second.

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by gemu2015** [comment]
the epoch number is too large to fit into float with second resolution. (all numbers are float in scripts)
use upsecs instead
if you need an absolute second resolution you may modify 
#define EPOCH_OFFSET 1546300800
in scripter source code to get a number that will fit into float
keep in mind that you only can hold 24 bits in second resolution => 16777215 seconds

**Reply by Andi-Coder** [reply]
It would be enough if the current epoch variable would be in an absolute second resolution. Is there any way to get this?

**Reply by barbudor** [reply]
https://github.com/arendst/Tasmota/blob/e7cbd4f00191025254ed9ff577a3f243a0c81c06/tasmota/xdrv_10_scripter.ino#L204

**Reply by barbudor** [reply]
If you change that to 1577836800 you will get the number of seconds since Jan 1st 2021
But anyhow you will reach the 24 bits resolution in 194 days

**Reply by Andi-Coder** [reply]
Ok, thank you. My intended is to count seconds beyond a reboot. Thats why I need a second time fix point. Then it doesnt matter if the device reboots. Do have another approach for that?

**Reply by barbudor** [reply]
_"to count seconds beyond a reboot"_
this is exactly what @gemu2015 suggested in te 1st part of his answer with `upsecs`

![image](https://user-images.githubusercontent.com/2996042/110754728-69402600-8248-11eb-9c8c-568f2d9804ec.png)

**Reply by Andi-Coder** [reply]
I mean, script should count seconds over a reboot. Upsecs is null after reboot. Thats the reason I need a fix time point to messure the second.

**Reply by barbudor** [reply]
If you mean the time BETWEEN reboots between a 1 second accuracy then you'd better do that in the backend from the MQTT messages published at boot.

**Reply by gemu2015** [reply]
ok i introduced a variable epoffs to which you can write the desired offset. however since this is also a float you can not set the start to an accurately selected point.
but then you get at least for 194 days a seconds resolution of epoch.

e.g. put this into >B
epoffs=1615461632

will do a pr soon
